src/linux/system-linux.c

```diff
@@ -151,12 +158,156 @@ frida_system_kill (guint pid)
 gchar *
 frida_temporary_directory_get_system_tmp (void)
 {
+  const gchar * tmp_dir;
+
 #ifdef HAVE_ANDROID
   if (getuid () == 0)
     return g_strdup ("/data/local/tmp");
 #endif
 
-  return g_strdup (g_get_tmp_dir ());
+  tmp_dir = g_get_tmp_dir ();
+
+  /*
+   * If the temporary directory resides on a file-system which is marked
+   * `noexec`, then we won't be able to write the `frida-agent.so` there and
+   * subsequently `dlopen` it inside the target application as it will result in
+   * permission denied.
+   *
+   * The mounting of the temporary file-system as `noexec` is sometimes used as
+   * an added security measure on embedded systems where the functionality is
+   * fixed and we aren't expecting any interactive user sessions.
+   *
+   * Since our current process is executing, we know that it must reside on a
+   * file-system which is not mounted `noexec`. Whilst it is possible that it is
+   * mounted read-only, or there may be some other reason why it isn't suitable,
+   * we know that the temporary directory is definitely unusable. If both these
+   * locations are found to be unsuitable, then a future implementation may seek
+   * to validate an ordered list of potential locations.
+   */
+  if (frida_is_directory_noexec (tmp_dir))
+  {
+    return get_application_directory ();
+  }
+  else
+  {
+    return g_strdup (tmp_dir);
+  }
+}
+
+static gboolean
+frida_is_directory_noexec (const gchar * directory)
+{
+  struct stat directory_stat, current_stat;
+  FILE *fp = NULL;
+  struct mntent mntent, *current;
+
+  /*
+   * Since the mounting of the temporary filesystem as `noexec` is not that
+   * common, we assume that this isn't the case unless we can determine
+   * otherwise definitively. (e.g. any errors whilst determining the status
+   * result in us assuming that the directory is executable).
+   */
+  gboolean noexec = FALSE;
+
+  /*
+   * The implementation of this function is derived from `find_mount_point` in
+   * busybox. It finds the mounted filesystem on which a given directory resides
+   * and checks the mount options for `noexec`.
+   */
+  if (stat(directory, &directory_stat) != 0)
+    goto beach;
+
+  if (!S_ISDIR(directory_stat.st_mode))
+    goto beach;
+
+  fp = setmntent ("/proc/mounts", "r");
+  if (fp == NULL)
+    goto beach;
+
+  for (
+    current = getmntent_r(fp, &mntent, mntent_buf, sizeof (mntent_buf));
+    current != NULL;
+    current = getmntent_r(fp, &mntent, mntent_buf, sizeof (mntent_buf)))
+  {
+    /*
+     * In Linux 2.6 rootfs mount always exists, and it makes sense to always
+     * ignore it.
+     */
+    if (g_strcmp0(current->mnt_fsname, "rootfs") == 0)
```

Comment:
(Here and elsewhere.) Can `current->mnt_fsname` be `NULL`? If not, we should be using `strcmp()`.

Reply:
The `man` page for `getmntent_r` doesn't explicitly say that the structure member can be NULL, or that it can't. Given we may find ourselves running on many different versions of `libc` and perhaps different variants too (e.g. uClibc, musl) is it prudent to be safe?

Reply:
Just checked the sources of glibc and uClibc, including an ancient glibc version, and determined that the worst-case is an empty string. So it seems safe to assume these are never `NULL`. I'll change these to `strcmp()` to make the code a bit clearer.